spec.html

```diff
@@ -14653,28 +14635,8 @@ <h1>The Addition Operator ( `+` )</h1>
         <h1>Runtime Semantics: Evaluation</h1>
         <emu-grammar>AdditiveExpression : AdditiveExpression `+` MultiplicativeExpression</emu-grammar>
         <emu-alg>
-          1. Let _lref_ be the result of evaluating |AdditiveExpression|.
-          1. Let _lval_ be ? GetValue(_lref_).
-          1. Let _rref_ be the result of evaluating |MultiplicativeExpression|.
-          1. Let _rval_ be ? GetValue(_rref_).
-          1. Let _lprim_ be ? ToPrimitive(_lval_).
-          1. Let _rprim_ be ? ToPrimitive(_rval_).
-          1. If Type(_lprim_) is String or Type(_rprim_) is String, then
-            1. Let _lstr_ be ? ToString(_lprim_).
-            1. Let _rstr_ be ? ToString(_rprim_).
-            1. Return the string-concatenation of _lstr_ and _rstr_.
-          1. Let _lnum_ be ? ToNumeric(_lprim_).
-          1. Let _rnum_ be ? ToNumeric(_rprim_).
-          1. If Type(_lnum_) is different from Type(_rnum_), throw a *TypeError* exception.
-          1. Let _T_ be Type(_lnum_).
-          1. Return _T_::add(_lnum_, _rnum_).
+          1. Return ? EvaluateNumericBinaryExpression(|AdditiveExpression|, `+`, |MultiplicativeExpression|).
```

Comment:
I prefer that we leave the string concat case of `+` inline in the Evaluation of AdditiveExpression and leave EvaluateNumericBinaryExpression to be entirely numeric.

Reply:
Oh, very nice idea.

Reply:
That would mean we have to leave this whole algorithm unchanged though, right?

Reply:
Oh, good point, EvaluateNumericBinaryExpression does the GetValue on the operands, so that can't be used. I think this Evaluation would need to inline the GetValue parts, the string-concat parts, and then call ApplyNumericBinaryOperator for the numeric `+` case. It's not as clean a separation as I originally envisioned, but I think I still like it better than having to chase down to ApplyNumericBinaryOperator to realize there's a string coercion case.

Reply:
If we hoist the string-concatenation semantics out of ApplyNumericBinaryOperator (ANBO), then those semantics would have to be handled not only in the Evaluation semantics for AdditiveExpression (as suggested above) but also in the Evaluation semantics for  AssignmentExpression. And if you DRY those up, you'd probably get a wrapper around ANBO. That is, the semantics of the wrapper operation would be the same as the ANBO proposed in this PR, just under a different name. So it'd be simpler to just figure out what that name would be, and rename this PR's ANBO to that. Here are some possibilities:
- ApplyNonBooleanlBinaryOperator
- ApplyStringOrNumericBinaryOperator
- ApplyBinaryOperatorForWhichAnAssignmentOperatorExists

(If we expanded to handle equality and relational operators, then we could call it ApplyStrictBinaryOperator, "strict" in the sense of always evaluating both operands.)

Reply:
Ah, I see what you mean with the @= templating in AssignmentExpression. Hm, that's a bit unfortunate. I can live with ApplyStringOrNumericBinaryOperator.